src/dune/file_tree.ml

```diff
@@ -63,20 +62,30 @@ module Dune_file = struct
     | Plain x -> x.path
     | Ocaml_script p -> p
 
-  let load file ~project =
-    Io.with_lexbuf_from_file (Path.source file) ~f:(fun lb ->
-        if Dune_lexer.is_script lb then
-          Ocaml_script file
-        else
-          let sexps = Dune_lang.Parser.parse lb ~mode:Many in
-          let decoder =
-            Dune_project.set_parsing_context project Sub_dirs.decode
-          in
-          let sub_dirs, sexps =
-            Dune_lang.Decoder.parse decoder Univ_map.empty
-              (Dune_lang.Ast.List (Loc.none, sexps))
-          in
-          Plain { path = file; sexps; sub_dirs })
+  let load_plain file sexps ~parent ~project =
+    let decoder = Dune_project.set_parsing_context project Sub_dirs.decode in
+    let active =
+      let parsed =
+        Dune_lang.Decoder.parse decoder Univ_map.empty
+          (Dune_lang.Ast.List (Loc.none, sexps))
+      in
+      match parent with
+      | None -> parsed
+      | Some parent -> Sub_dirs.Dir_map.merge parsed parent
+    in
+    let contents = Sub_dirs.Dir_map.root active in
+    { Plain.path = file; contents; for_subdirs = active }
+
+  let load file ~file_exists ~parent ~project =
+    match file_exists with
+    | false -> Plain (load_plain file [] ~parent ~project)
+    | true ->
+      Io.with_lexbuf_from_file (Path.source file) ~f:(fun lb ->
+          if Dune_lexer.is_script lb then
+            Ocaml_script file
```

Comment:
So here we can add a check that `parent` is "trivial" and then there's no obscure issue anymore, just an explicit error.

Reply:
Actually, now I'm wondering if we should just fix the behavior. Tbh, it really shouldn't be hard  to support. We just need to extend representation from:

```ocaml
type t =
  | Plain of Plain.t
  | Ocaml_script of Path.Source.t
```

to:

```ocaml
type t =
  | Plain of Plain.t
  | Ocaml_script of { plain : Plain.t ; source : Path.Source.t }
```

Reply:
Fair enough, I think that should work. I would name the `plain` field `from_parent` for consistency with the other place.

Reply:
The last commit implements this. It seems to work fine but reveals a lot of duplication between `File_tree.Dune_file` and `Dune_load.Dune_file.`. I think these two modules should be unified. Anyway, it's not a trivial task so I will save it for another PR if you don't mind.